dev/com.ibm.ws.security.common.jsonwebkey/src/com/ibm/ws/security/common/jwk/impl/JwKRetriever.java

```diff
@@ -0,0 +1,513 @@
+		this.jwkSet = jwkSet; // get the JWKSet from the Config
+		this.sslSupport = sslSupport;
+		this.hostNameVerificationEnabled = hnvEnabled;
+	}
+
+//	public JwKRetriever(JwtConsumerConfig config) {
+//		configId = config.getId();
+//		sslConfigurationName = config.getSslRef();
+//		jwkEndpointUrl = config.getJwkEndpointUrl();
+//		jwkSet = config.getJwkSet();
+//		hostNameVerificationEnabled = config.isHostNameVerificationEnabled();
+//		
+//	}
+
+	/**
+	 * Either kid or x5t will work. But not both
+	 *
+	 * @param kid
+	 * @param x5t
+	 * @return
+	 * @throws PrivilegedActionException
+	 * @throws IOException
+	 * @throws KeyStoreException
+	 * @throws InterruptedException
+	 */
+	@FFDCIgnore({ KeyStoreException.class })
+	public PublicKey getPublicKeyFromJwk(String kid, String x5t)
+			throws PrivilegedActionException, IOException, KeyStoreException, InterruptedException {
+		PublicKey key = this.getJwkCache(kid, x5t);
+		KeyStoreException errKeyStoreException = null;
+		InterruptedException errInterruptedException = null;
+		if (key == null) {
+			try {
+				key = this.getJwkRemote(kid, x5t);
+			} catch (KeyStoreException e) {
+				errKeyStoreException = e;
+			} catch (InterruptedException e) {
+				errInterruptedException = e;
+			}
+		}
+		if (key == null) {
+			if (errKeyStoreException != null) {
+				throw errKeyStoreException;
+			}
+			if (errInterruptedException != null) {
+				throw errInterruptedException;
+			}
+		}
+		return key;
+	}
+
+	protected PublicKey getJwkCache(String kid, String x5t) {
+		if (kid != null) {
+			return jwkSet.getPublicKeyByKid(kid);
+		} else if (x5t != null) {
+			return jwkSet.getPublicKeyByx5t(x5t);
+		}
+		return jwkSet.getPublicKeyByKid(null);
+	}
+
+	protected PublicKey getJwkLocal(String kid, String x5t) throws PrivilegedActionException, IOException {
+		// JWKSet jwkset = clientConfig.getJwkSet();
+		try {
+			final String keyFile = keyFileName;// clientConfig.getJsonWebKey();
+			if (keyFile == null) {
+				return null;
+			}
+
+			FileInputStream inputStream = (FileInputStream) AccessController
+					.doPrivileged(new PrivilegedExceptionAction<Object>() {
+						@Override
+						public Object run() throws Exception {
+							File fileJwk = new File(keyFile);
+							if (fileJwk.exists()) {
+								return new FileInputStream(fileJwk);
+							} else {
+								return null;
+							}
+						}
+					});
+			if (inputStream == null) {
+				return null;
+			}
+			parseJwk(null, inputStream, jwkSet, sigAlg);
+
+			// Object obj = parser.parse(fileReader);
+
+			// JsonObject jsonObject = (JsonObject) obj;
+			// for (JsonElement element : jsonObject.getAsJsonArray("keys")) {
+			// JsonObject keyObject = element.getAsJsonObject();
+			// JWK jwk = Jose4jRsaJWK.getInstance(keyObject);
+			// if (jwk != null) {
+			// jwk.parse();
+			// jwkset.addJWK(jwk);
+			// if (tc.isDebugEnabled()) {
+			// Tr.debug(tc, "add local key for keyid: ", jwk.getKeyID());
+			// }
+			// }
```

Comment:
removing the comments might be good

Reply:
Sure. There will be 2-3 rounds of deliveries in this code path to support common JWK impl.